### PR TITLE
Throw exception on returning a unique_ptr or shared_ptr nullptr (or any other holder type) from py::init, rather than crashing

### PR DESCRIPTION
**Pull request by YannickJadoul**
One overload of the implementation of `py::init` was missing this check for a null pointer that's present in the other two. This should fix it.

Crash found by @rgkw in https://github.com/pybind/pybind11/pull/2152#issuecomment-678816799.

### CODE REVIEW

**Comment by YannickJadoul** [comment]
> Thanks Yannick for jumping in plugging this hole!

Got curious ;-)

Let's get one more review (@henriii, @bstaletic, or maybe @EricCousineau-TRI? Don't know who's familiar with this part of the code). Then we should be fine, given how minor this change is :-)

**Review by bstaletic — APPROVED** [review]
Looks fine to me.

**Comment by YannickJadoul** [comment]
Ow, I had tagged the wrong @henryiii, in the [comment before](https://github.com/pybind/pybind11/pull/2430#issuecomment-679043209). But given @bstaletic's review, I'll merge tomorrow morning, if no one speaks up :-)

**Comment by YannickJadoul** [comment]
OK, merging :-) If there's still a problem we overlooked, it can be fixed later.

**Pull request merged**
